### ISSUE TITLE
Cannot set route on Raspbian with Rpi

### ISSUE THREAD

**Issue by jbremond** [adrienverge/openfortivpn#426]
Hi,

I have installed the latest tagged version 1.8.1 on a RPI 3, distribution Raspbian 9.6.
When connecting to my Fortinet SSL Gateway, it seems that openfortivpn cannot set the route.

```
INFO:   negotiation complete
DEBUG:  pppd ---> gateway (6 bytes)
DEBUG:  Got Address: 10.212.134.200
DEBUG:  Interface Name: ppp0
DEBUG:  Interface Addr: 10.212.134.200
INFO:   Interface ppp0 is UP.
INFO:   Setting new routes...
DEBUG:  ip route show to 0.0.0.0/0.0.0.0 dev !ppp0
DEBUG:  ip route show to 192.168.10.14/255.255.255.255 dev ppp0
DEBUG:  Route not found.
DEBUG:  ip route show to 192.168.10.14/255.255.255.255 dev !ppp0
DEBUG:  Setting route to vpn server...
DEBUG:  ip route show to 192.168.10.14/255.255.255.255 via 255.255.255.127 dev enxb827ebdb08e2
DEBUG:  ip route add to 192.168.10.14/255.255.255.255 via 255.255.255.127 dev enxb827ebdb08e2
WARN:   Could not set route to vpn server (Network is unreachable).
DEBUG:  Deleting the current default route...
DEBUG:  ip route del to 0.0.0.0/0.0.0.0 via 255.255.255.127 dev enxb827ebdb08e2
WARN:   Could not delete the current default route (No such process).
DEBUG:  Setting new default route...
DEBUG:  ip route add to 0.0.0.0/0.0.0.0 dev ppp0

```

What is strange is the gateway address that openfortivpn want to add:
`ip route show to 192.168.10.14/255.255.255.255 via 255.255.255.127 dev enxb827ebdb08e2`

255.255.255.127 is not a gateway.

I don't know from where this gateway is calculated, but the problems seems coming from there. Any ideas ?

Thanks,
Joris

**Comment by DimitriPapadopoulos** [comment]
Note sure why it happens, but the relevant messages are printed in _ipv4_protect_tunnel_route_():
https://github.com/adrienverge/openfortivpn/blob/7af5394/src/ipv4.c#L687-L715
The _tunnel_ argument passed to _ipv4_protect_tunnel_route_() might already be wrong. Unless someone has a better idea, you could add some extra debugging code and re-compile. Print the contents of _tunnel_ and other variables in _ipv4_protect_tunnel_route_().

**Comment by DimitriPapadopoulos** [comment]
Perhaps it would help if you could show us the routes before/after starting _openfortivpn_.

**Comment by jbremond** [comment]
I have added 3 debugs logs to print the gateway value of the tunnel. You will find the value in BOLD.
Also, information to add is the tunnel is up and running, only the route are not set.

```
INFO:   negotiation complete
DEBUG:  pppd ---> gateway (6 bytes)
DEBUG:  Got Address: 10.212.134.200
DEBUG:  Interface Name: ppp0
DEBUG:  Interface Addr: 10.212.134.200
INFO:   Interface ppp0 is UP.
INFO:   Setting new routes...
DEBUG:  ip route show to 0.0.0.0/0.0.0.0 dev !ppp0
```
**DEBUG:  tunnel->config->gateway_ip.s_addr == 235579584**
**DEBUG:  tunnel->ppp_iface == ppp0**
```
DEBUG:  ip route show to 192.168.10.14/255.255.255.255 dev ppp0
DEBUG:  Route not found.
DEBUG:  ip route show to 192.168.10.14/255.255.255.255 dev !ppp0
```
**DEBUG:  route_dest(gtw_rt).s_addr == 235579584**
```
DEBUG:  Setting route to vpn server...
DEBUG:  ip route show to 192.168.10.14/255.255.255.255 via 255.255.255.127 dev enxb827ebdb08e2
DEBUG:  ip route add to 192.168.10.14/255.255.255.255 via 255.255.255.127 dev enxb827ebdb08e2
WARN:   Could not set route to vpn server (Network is unreachable).
```

235579584 == 14.10.168.192 that correspond to my gateway: 192.168.10.14
I don't know if it is normal that this is print reversed ?

Bellow, the route on my RPI before launching the openfortivpn:
```
root@raspberrypi:/home/pi/openfortivpn# route -n
Table de routage IP du noyau
Destination     Passerelle      Genmask         Indic Metric Ref    Use Iface
0.0.0.0         192.168.20.254  0.0.0.0         UG    202    0        0 enxb827ebdb08e2
192.168.20.0    0.0.0.0         255.255.255.0   U     202    0        0 enxb827ebdb08e2
```
Nothing specific to my gateway: 192.168.10.14.

after setting the tunnel:
```
root@raspberrypi:/home/pi/openfortivpn# route -n
Table de routage IP du noyau
Destination     Passerelle      Genmask         Indic Metric Ref    Use Iface
0.0.0.0         0.0.0.0         0.0.0.0         U     0      0        0 ppp0
0.0.0.0         192.168.20.254  0.0.0.0         UG    202    0        0 enxb827ebdb08e2
192.0.2.1       0.0.0.0         255.255.255.255 UH    0      0        0 ppp0
192.168.20.0    0.0.0.0         255.255.255.0   U     202    0        0 enxb827ebdb08e2
```

**Comment by jbremond** [comment]
I have also print some variable inside ipv4_set_route

```
static int ipv4_set_route(struct rtentry *route)
{
#ifdef HAVE_RT_ENTRY_WITH_RT_DST
        /* we can copy rtentry struct directly between openfortivpn and kernel */
        log_debug("ip route add %s\n", ipv4_show_route(route));

        struct sockaddr_in *ipv4 = (struct sockaddr_in *)route->rt_gateway.sa_data;
        char ipAddress[INET_ADDRSTRLEN];
        inet_ntop(AF_INET, &(ipv4->sin_addr), ipAddress, INET_ADDRSTRLEN);
        printf("The IP address is: %s\n", ipAddress);
        printf("route->rt_gateway.sa_data %s\n",route->rt_gateway.sa_data);

```

and I get the following result:

```
INFO:   Setting new routes...
DEBUG:  ip route show to 0.0.0.0/0.0.0.0 dev !ppp0
DEBUG:  tunnel->config->gateway_ip.s_addr == 235579584
DEBUG:  tunnel->ppp_iface == ppp0
DEBUG:  ip route show to 192.168.10.14/255.255.255.255 dev ppp0
DEBUG:  Route not found.
DEBUG:  ip route show to 192.168.10.14/255.255.255.255 dev !ppp0
DEBUG:  route_dest(gtw_rt).s_addr == 235579584
DEBUG:  Setting route to vpn server...
DEBUG:  ip route show to 192.168.10.14/255.255.255.255 via 255.255.255.127 dev enxb827ebdb08e2
DEBUG:  ip route add to 192.168.10.14/255.255.255.255 via 255.255.255.127 dev enxb827ebdb08e2
The IP address is: 255.127.0.0
route->rt_gateway.sa_data
```

Ip address seems strange !

**Comment by mrbaseman** [comment]
Hi @jbremond 

> What is strange is the gateway address that openfortivpn want to add:
> > ip route show to 192.168.10.14/255.255.255.255 via 255.255.255.127 dev enxb827ebdb08e2
> > 255.255.255.127 is not a gateway.

right. That's unexpected. It could be something special on Raspbian. We use uint_32 at some places, but if the compiler ignores the request for **unsigned** and uses a signed 32 bit int internally, it could perhaps explain such an effect. Sure, the value is not a valid gateway, but it is something different from 255.255.255.255. 
It could as well be something in handling the uint_32 values in openfortivpn (an implicit conversion which has been working on other platforms so far...)

> 235579584 == 14.10.168.192 that correspond to my gateway: 192.168.10.14
> I don't know if it is normal that this is print reversed ?

that's expected. It's [network byte order](https://www.gnu.org/software/libc/manual/html_node/Byte-Order.html)

> after setting the tunnel:
> 
> > root@raspberrypi:/home/pi/openfortivpn# route -n
> > Table de routage IP du noyau
> > Destination     Passerelle      Genmask         Indic Metric Ref    Use Iface
> > 0.0.0.0         0.0.0.0         0.0.0.0         U     0      0        0 ppp0
> > 0.0.0.0         192.168.20.254  0.0.0.0         UG    202    0        0 enxb827ebdb08e2
> > 192.0.2.1       0.0.0.0         255.255.255.255 UH    0      0        0 ppp0
> > 192.168.20.0    0.0.0.0         255.255.255.0   U     202    0        0 enxb827ebdb08e2

it should remove the default route over enxb827ebdb08e2 and add a host route to the vpn server.
Perhaps the "--half-internet-routes=1" option works for you (as a workaround). 

> > DEBUG:  ip route add to 192.168.10.14/255.255.255.255 via 255.255.255.127 dev enxb827ebdb08e2
> > The IP address is: 255.127.0.0
> > route->rt_gateway.sa_data
> 
> Ip address seems strange !

the debug output looks quite ok, except for the gateway. I think something with your additional debugging code is weird. I think we should focus on the gateway value. 
Do you have `/proc/net/route` on your system? How do the routes look there before and after starting _openfortivpn_?

**Comment by jbremond** [comment]
Hi @mrbaseman ,
Thank you for your explanation. Yes I really think the gateway value is not good.
To answer your questions:
* Yes I have `/proc/net/route`
* The workaround using `--half-internet-routes=1` does not work (the program still try to set the route with "bad" value)
* regarding route before/after, I already put it in previous comment: https://github.com/adrienverge/openfortivpn/issues/426#issuecomment-460008010

Do you have an idea regarding uint_32 that I can check to validate your feeling ?

**Comment by DimitriPapadopoulos** [comment]
If I'm not mistaken the gateway IP address value is calculated here:
https://github.com/adrienverge/openfortivpn/blob/7af5394/src/tunnel.c#L388-L410

Can you investigate the values of `tunnel->config->gateway_ip = ((struct sockaddr_in *)result->ai_addr)->sin_addr;` and `inet_ntoa(tunnel->config->gateway_ip)`?

**Comment by mrbaseman** [comment]
Hi @jbremond,

could you show the output of `cat /proc/net/route`, please? That's what _opnefortivpn_ tries to parse when that file exists. In that case it does not parse the output of the `route` command. Your interface name is quite long, so something might go wrong here (e.g.  interface and destination glued together in the output and therefore the columns not parsed correctly).

**Comment by jbremond** [comment]
Hello @mrbaseman,

The output is the following:

`cat /proc/net/route`


```
Iface   Destination     Gateway         Flags   RefCnt  Use     Metric  Mask            MTU     Window  IRTT
enxb827ebdb08e2 00000000        FE14A8C0        0003    0       0       202     00000000        0       0       0
enxb827ebdb08e2 0014A8C0        00000000        0001    0       0       202     00FFFFFF        0       0       0

```

**Comment by jbremond** [comment]
@mrbaseman to investigate what you think, I have created an udev rule to have a short interface name:

`SUBSYSTEM=="net", ACTION=="add", DRIVERS=="?*", ATTR{address}=="xx:xx:xx:xx:xx:xx", ATTR{dev_id}=="0x0", ATTR{type}=="1", KERNEL=="eth*", NAME="eth0"`

When I start the VPN, I still get the same error:
```
DEBUG:  ip route show to 192.168.10.14/255.255.255.255 dev !ppp0
DEBUG:  Setting route to vpn server...
DEBUG:  ip route show to 192.168.10.14/255.255.255.255 via 255.255.255.127 dev eth0
DEBUG:  ip route add to 192.168.10.14/255.255.255.255 via 255.255.255.127 dev eth0
WARN:   Could not set route to vpn server (Network is unreachable).
DEBUG:  Deleting the current default route...
DEBUG:  ip route del to 0.0.0.0/0.0.0.0 via 255.255.255.127 dev eth0
WARN:   Could not delete the current default route (No such process).

```

**Comment by jbremond** [comment]
@DimitriPapadopoulos the gateway value seems good:

`printf("GATEWAY_IP=%s\n",inet_ntoa(tunnel->config->gateway_ip));`

`GATEWAY_IP=192.168.10.14`

**Comment by mrbaseman** [comment]
I believe that I have found something, now: [ipv4_get_route](https://github.com/adrienverge/openfortivpn/blob/master/src/ipv4.c#L349-L359) uses **strtol** to convert the parsed strings to long and assigns those to unsigned int. 

On Raspbian however, both, [long and int are 32 bit](https://raspberry-projects.com/pi/programming-in-c/memory/variables). So, a signed long does not cover the full range of an unsigned int.

@jbremond could you test my [raspi branch](https://github.com/mrbaseman/openfortivpn/tree/raspi) please? I use strtoul there, which should hopefully solve the problem.

**Comment by jbremond** [comment]
@mrbaseman I just test the code in your branch and it works, like a charm !
Thank for your help and the explanation of the reason !

**Comment by mrbaseman** [comment]
@jbremond thanks for testing and for the feedback. I have just merged these changes into the master branch. They will be in the next release.

**Comment by mrbaseman** [comment]
The fix went into the [1.9.0](https://github.com/adrienverge/openfortivpn/releases/tag/v1.9.0) release
